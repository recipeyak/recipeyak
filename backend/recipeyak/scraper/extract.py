"""
IO-free recipe-scraping
"""

from __future__ import annotations

import base64
import binascii
import json
from dataclasses import dataclass
from typing import cast

import extruct
from markdownify import markdownify
from recipe_scrapers import AbstractScraper
from recipe_scrapers._exceptions import SchemaOrgException

from recipeyak.scraper.format_time import human_time_duration
from recipeyak.string import starts_with


@dataclass(frozen=True, slots=True, kw_only=True)
class _ExtractedRecipe:
    title: str | None
    total_time: str | None
    yields: str | None
    instructions: list[str]
    # the first param of the tuple will be None in the case of a
    ingredient_groups: list[_IngredientGroup]
    author: str | None
    canonical_url: str | None
    image_urls: list[str | None]


@dataclass(frozen=True, slots=True, kw_only=True)
class _IngredientGroup:
    name: str | None
    ingredients: list[str]


def _extract_open_graph_image(html: bytes | str) -> str | None:
    og_data = extruct.extract(html, syntaxes=["opengraph"], uniform=True)["opengraph"]
    if not og_data:
        return None
    return cast(str | None, og_data[0].get("og:image"))


def _extract_tips(parsed: AbstractScraper) -> list[str]:
    nextjs_data_tag = parsed.soup.find(
        "script", type="application/json", id="__NEXT_DATA__"
    )
    if nextjs_data_tag is None:
        return []
    # TODO: make more robust, this is really only intended for nyt cooking
    try:
        json_next_data = json.loads(base64.b64decode(nextjs_data_tag.text))
    except (json.JSONDecodeError, binascii.Error):
        return []
    try:
        tips: list[str] = json_next_data["props"]["pageProps"]["recipe"]["tips"]
    except KeyError:
        # https://www.americastestkitchen.com/recipes/12373-apple-fennel-remoulade
        # uses nextjs but doesn't have the same structure as nyt cooking for
        # their props ofc
        return []
    return tips


def _extract_total_time(parsed: AbstractScraper) -> str | None:
    try:
        total_time = parsed.total_time()
        if total_time is None:
            return None
        return human_time_duration(total_time * 60)
    except SchemaOrgException:
        return None


_YIELD_PREFIXES = {"Serves", "Servings", "Makes", "Creates"}


def _extract_yields(parsed: AbstractScraper) -> str | None:
    # We don't want to use the library's yields method as it's buggy
    # see: https://github.com/hhursev/recipe-scrapers/issues/960
    if yield_data := parsed.schema.data.get("recipeYield") or parsed.schema.data.get(
        "yield"
    ):
        if isinstance(yield_data, list):
            yield_data = yield_data[0]
        yield_str = str(yield_data)
        for prefix in _YIELD_PREFIXES:
            if starts_with(yield_str, prefix):
                return yield_str.removeprefix(prefix)
            return yield_str
    return None


def _extract_author(parsed: AbstractScraper) -> str | None:
    try:
        return cast(str | None, parsed.author())
    except AttributeError:
        return None


def _extract_ingredient_groups(parsed: AbstractScraper) -> list[_IngredientGroup]:
    try:
        ingredient_groups = list[_IngredientGroup]()
        for group in parsed.ingredient_groups():
            ingredient_groups.append(
                _IngredientGroup(name=group.purpose, ingredients=group.ingredients)
            )
        return ingredient_groups
    except ValueError:
        # There's a chance the library will throw an error
        return [_IngredientGroup(name=None, ingredients=parsed.ingredients())]


def extract_recipe(parsed: AbstractScraper) -> _ExtractedRecipe:
    """
    Extact what data we can from the html without doing any IO

    There is some IO involved in elsewhere for determining the best image to
    use, but this module should be io-free, which makes testing and structuring
    easier
    """
    total_time = _extract_total_time(parsed)
    tips_html = _extract_tips(parsed)
    yields = _extract_yields(parsed)
    author = _extract_author(parsed)
    ingredient_groups = _extract_ingredient_groups(parsed)
    canonical_url = parsed.canonical_url()
    instructions = parsed.instructions_list()
    if tips_html:
        # NOTE: we don't have a tips construct in the data model so we stuff
        # them into a final step and add a markdown header
        tips_markdown = [markdownify(t) for t in tips_html]
        tips_blob = "\n\n".join(tips_markdown)
        instructions.append("**Tips**\n" + tips_blob)
    title = parsed.title()

    # possible image urls that we figure out the best one later on by actually fetching them
    og_image_url = _extract_open_graph_image(parsed.page_data)
    image_urls = [og_image_url, parsed.image()]

    return _ExtractedRecipe(
        canonical_url=canonical_url,
        title=title,
        total_time=total_time,
        ingredient_groups=ingredient_groups,
        yields=yields,
        author=author,
        instructions=instructions,
        image_urls=image_urls,
    )
