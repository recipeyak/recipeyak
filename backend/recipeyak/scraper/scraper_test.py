from pathlib import Path

from syrupy.assertion import SnapshotAssertion

from recipeyak.scraper.scrape_recipe import _parse_recipe


def test_parse_recipe_tips_html_to_markdown(
    snapshot: SnapshotAssertion,
) -> None:
    """
    Test with __NEXT_DATA__ being base64 encoded.

    Check that we convert tips from html to markdown.
    """
    html = (
        Path(__file__).parent
        / "test_data"
        / "1023609-chile-crisp-fettuccine-alfredo-with-spinach.html"
    ).read_bytes()

    result = _parse_recipe(
        html=html,
        url="https://cooking.nytimes.com/recipes/1023609-chile-crisp-fettuccine-alfredo-with-spinach",
    )
    assert result == snapshot()


def test_parse_recipe_tips_html_to_markdown_plain_text(
    snapshot: SnapshotAssertion,
) -> None:
    """
    Check that we convert tips from html to markdown.

    Test with __NEXT_DATA__ being plain text.
    """
    html = (
        Path(__file__).parent / "test_data" / "1025652-oven-smores.html"
    ).read_bytes()

    result = _parse_recipe(
        html=html,
        url="https://cooking.nytimes.com/recipes/1025652-oven-smores",
    )
    assert result == snapshot()


def test_parse_links_in_steps(snapshot: SnapshotAssertion) -> None:
    """
    Check that we convert links from html to markdown in steps.
    """
    html = (
        Path(__file__).parent
        / "test_data"
        / "pressure-cooker-fast-and-easy-chicken-enchiladas-food-lab-recipe.html"
    ).read_bytes()

    result = _parse_recipe(
        html=html,
        url="https://www.seriouseats.com/pressure-cooker-fast-and-easy-chicken-enchiladas-food-lab-recipe",
    )
    assert result == snapshot()
