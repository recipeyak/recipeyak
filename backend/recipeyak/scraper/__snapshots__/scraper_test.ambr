# name: test_parse_links_in_steps
  _ExtractedRecipe(title='Easy Pressure Cooker Chicken Enchiladas Recipe', total_time='1 hour, 5 mins', yields='4', instructions=['For the Chicken and Sauce: Trim tops of chiles and discard seeds (see here for more detailed instructions ). Place on a microwave-safe plate and microwave on high power until pliable and fragrant, about 15 seconds. Cut chiles into thin strips using kitchen shears or a sharp knife.', 'Combine chiles, chicken, serrano or jalapeño, tomatoes, onion, garlic, chipotle peppers and adobo sauce, bay leaf, oregano, cumin, coriander, and chicken stock in a pressure cooker, season with salt and pepper, and stir roughly until ingredients are evenly distributed. Seal pressure cooker and cook on high pressure for 15 minutes. Release pressure using quick-release method. Using tongs, transfer chicken to a bowl and set aside to cool.', 'Discard bay leaf. Transfer remaining ingredients from pressure cooker to the jar of a blender. Add soy sauce and lime juice. Blend, starting at low speed and slowly increasing to high until sauce is smooth. Season to taste with salt and stir in 3 tablespoons of cilantro. Transfer to a bowl.', 'When chicken is cool enough to handle, discard skin and bones. Shred chicken meat roughly with your fingers or two forks. Fold in 1/4 cup of sauce and season to taste with salt and pepper.', 'To Assemble: Adjust oven rack to middle position and preheat oven to 400°F (200°C). Heat canola oil in a medium cast iron or non-stick skillet over medium heat until shimmering. Working one at a time, place a tortilla in the oil, fry for 5 seconds, turn, and fry second side for 5 seconds. Transfer to a paper towel-lined plate to drain, then to a second clean plate. Stack tortillas as you work and cover with a clean kitchen towel or paper towel when finished to keep tortillas warm.', 'Dip tortillas in sauce one at a time, letting excess drip off and stacking them as you go. Ladle one cup of sauce over the base of a 9- by 13-inch baking dish. Place 2 to 3 tablespoons of chicken filling in a line along the center of one tortilla. Roll tortilla up like a cigar and transfer to baking dish, seam side-down. Repeat with remaining tortillas and filling.', 'Spoon more sauce over the top of tortillas, then spread cheddar or Jack cheese in two even rows across the center of tortillas. Spoon a few more tablespoons of sauce on top of cheese. Cover baking dish with aluminum foil, transfer to oven, and bake for 10 minutes. Remove foil, sprinkle with half of cotija cheese, and continue baking until heated through, 5 to 10 minutes longer.', 'Remove baking dish from oven, sprinkle with remaining cotija cheese and chopped cilantro, and serve immediately, passing any additional salsa at the table.'], ingredient_groups=[_IngredientGroup(name=None, ingredients=['For the Chicken and Sauce:', '2 whole dried ancho chilies', '2 pounds bone-in, skin-on chicken thighs', '1 Serrano or jalapeño pepper, roughly sliced', '1 pound roma tomatoes, roughly chopped (about 4)', '1 medium onion, sliced', '4 medium cloves garlic, smashed', '2 whole canned chipotle chilies packed in adobo, plus 2 tablepoons sauce from the can', '1 bay leaf', '2 teaspoons dried oregano, preferably Mexican', '1 teaspoon ground cumin seed', '1/2 teaspoon ground coriander seed', '1/2 cup homemade or store-bought low-sodium chicken stock', 'Kosher salt and freshly ground black pepper', '1 tablespoon juice from 1 lime', '1 tablespoon soy sauce', '1/4 cup chopped fresh cilantro leaves and fine stems', 'To Assemble:', '1/4 cup canola oil', '12 fresh corn tortillas', '3 ounces grated Jack or cheddar cheese', '1 ounce crumbled cotija cheese'])], author='J. Kenji López-Alt', canonical_url='https://www.seriouseats.com/pressure-cooker-fast-and-easy-chicken-enchiladas-food-lab-recipe', image_urls=['https://www.seriouseats.com/thmb/MEehZwvU0VrQiq6XNCsVqERrnig=/1500x0/filters:no_upscale():max_bytes(150000):strip_icc()/__opt__aboutcom__coeus__resources__content_migration__serious_eats__seriouseats.com__recipes__images__2015__04__20150416-pressure-cooke-chicken-enchiladas-recipe-28-6da39b7c0ff3429ab2d8e2ca927d8ba5.jpg', 'https://www.seriouseats.com/thmb/MEehZwvU0VrQiq6XNCsVqERrnig=/1500x0/filters:no_upscale():max_bytes(150000):strip_icc()/__opt__aboutcom__coeus__resources__content_migration__serious_eats__seriouseats.com__recipes__images__2015__04__20150416-pressure-cooke-chicken-enchiladas-recipe-28-6da39b7c0ff3429ab2d8e2ca927d8ba5.jpg'])
# ---
# name: test_parse_recipe_tips_html_to_markdown
  _ExtractedRecipe(title='Chile Crisp Fettuccine Alfredo With Spinach', total_time='25 mins', yields='6 servings', instructions=['Bring a large pot of salted water to a boil. While the water heats, melt the butter with the chile crisp in a very large skillet or Dutch oven over low heat.', 'Whisk in the cream and keep warm over low. (It should steam, not bubble.)', 'Cook the fettuccine until al dente according to the package directions. Use tongs to transfer the noodles to the cream mixture, reserving the pasta water.', 'Add the spinach and turn with tongs until the noodles are well coated.', 'Add the Parmesan and toss, still over low heat, until the noodles are slicked with a creamy sauce, adding a spoonful or two of pasta water if needed to loosen the sauce. Divide among serving dishes and top with Parmesan and more chile crisp, if you’d like. Serve immediately.', '**Tips**\nYou can make [chile crisp](https://cooking.nytimes.com/recipes/1022366-chile-crisp) easily at home or buy it in supermarkets or online. It varies in spiciness, so adjust the amount to your taste. For this dish, try to add more of the solids than the oil to the sauce for the most flavorful dish.'], ingredient_groups=[_IngredientGroup(name=None, ingredients=['Salt', '4 tablespoons butter', '1 to 2 tablespoons chile crisp, plus more to taste (see Tip)', '1 cup heavy cream', '1 pound dried fettuccine', '1 (5-ounce) package baby spinach', '3/4 cup finely grated Parmesan (2 1/4 ounces), plus more for serving'])], author='Genevieve Ko', canonical_url='https://cooking.nytimes.com/recipes/1023609-chile-crisp-fettuccine-alfredo-with-spinach', image_urls=['https://static01.nyt.com/images/2022/11/02/dining/27korex-alfredo/26korex-alfredo-threeByTwoMediumAt2X.jpg', 'https://static01.nyt.com/images/2022/11/02/dining/27korex-alfredo/26korex-alfredo-videoSixteenByNineJumbo1600.jpg'])
# ---
